### PR TITLE
Implement the import feature

### PR DESCRIPTION
**Pull request by kmizu**
- Now you can import other YAML config files.

### CODE REVIEW

**Comment by ybiquitous** [comment]
Thank you for the implementation!

I have a question. Can we detect a typo such as `impor:` in `javasee.yml`?
E.g.

```
$ cat javasee.yml
rules: []

impor: []

$ javasee test
Checking rule id uniqueness...
Checking rule patterns...
Tested 0 rules with 0 tests
  All tests green!
```

Hopefully expected output:

```
$ javasee test
YAML file has schema error: Unknown key: impor
```

**Comment by kmizu** [comment]
> Thank you for the implementation!
> 
> I have a question. Can we detect a typo such as `impor:` in `javasee.yml`?
> E.g.
> 
> ```
> $ cat javasee.yml
> rules: []
> 
> impor: []
> 
> $ javasee test
> Checking rule id uniqueness...
> Checking rule patterns...
> Tested 0 rules with 0 tests
>   All tests green!
> ```
> 
> Hopefully expected output:
> 
> ```
> $ javasee test
> YAML file has schema error: Unknown key: impor
> ```

I've fixed in https://github.com/sider/JavaSee/pull/104/commits/ee89e1910d0b0db8e0531ba02da238a77da8da5d :

image

**Comment by ybiquitous** [comment]
Thank you. I could confirm the fix. 👍 

BTW, is it acceptable `javasee.yml` has no `rules` and has only `import`? What do you think?

```
$ cat javasee.yml
import:
  - javasee2.yml

$ javasee test
YAML file has schema error: Missing key: rules
```

**Comment by kmizu** [comment]
> Thank you. I could confirm the fix. 👍
> 
> BTW, is it acceptable `javasee.yml` has no `rules` and has only `import`? What do you think?
> 
> ```
> $ cat javasee.yml
> import:
>   - javasee2.yml
> 
> $ javasee test
> YAML file has schema error: Missing key: rules
> ```

In current implementation, it's not acceptable.  However, the following code works:

```yaml
- rules: []
- import: javasee2.yml
```

I think that no `rules` should not be allowed since the writer could forget to write `rules`.  Even if it is required, empty `rules` works.

**Comment by ybiquitous** [comment]
Make sense. 👍

**Comment by ybiquitous** [comment]
Does it make sense to return `CONFIG_FILE_NOT_FOUND (3)` or some meaningful exit code if importing file is not found?

Here is the current behavior:

```
$ cat javasee.yml
rules: []

import:
  - not_found.yml

$ javasee test
There is nothing to test at not_found.yml ...
Make a configuration and run test again!
com.github.sider.javasee.Exceptions$ConfigFileException
	at com.github.sider.javasee.Config.readFrom(Config.java:29)
	at com.github.sider.javasee.Config.init(Config.java:56)
	at com.github.sider.javasee.Config.readFrom(Config.java:32)
	at com.github.sider.javasee.Config.load(Config.java:80)
	at com.github.sider.javasee.command.TestCommand.start(TestCommand.java:66)
	at com.github.sider.javasee.Main.lambda$main$2(Main.java:61)
	at java.base/java.util.Optional.map(Optional.java:265)
	at com.github.sider.javasee.Main.main(Main.java:61)
```

I think the following behavior is more useful:

```
$ javasee test
not_found.yml being imported is not found ...

$ echo $?
3
```

What do you think?

**Comment by kmizu** [comment]
OK.  I see it make sense.  It will be fixed later.

**Comment by kmizu** [comment]
@ybiquitous 

https://github.com/sider/JavaSee/pull/104#issuecomment-689979243 the feature is implemented

image

**Comment by ybiquitous** [comment]
Thanks, I could confirm it is fixed! 👍 

In addition, could you please improve the error message about the configuration file?
When using multiple configuration files, uses cannot know in which file an error occurs:

```
$ cat javasee.yml
rules:
  - id: a
    pattern: a
    message: a

import:
  - src/javasee2.yml

$ cat src/javasee2.yml
rules:
  - id: a
    pattern: a
    message_: a

$ javasee test
YAML file has schema error: Unknown keys: [message_], only [tests, id, justification, message, pattern] are allowed
```

Expected, for example:

```
"src/javasee2.yml" has schema error: Unknown keys: [message_], only [tests, id, justification, message, pattern] are allowed
```

**Comment by kmizu** [comment]
@ybiquitous 

The feature has been implemented in: https://github.com/sider/JavaSee/pull/104/commits/bf12ed4698d03745325e9f527ac1f78156f59c12

image

File: src/main/java/com/github/sider/javasee/command/CheckCommand.java [hunk 1]
```diff
@@ -47,30 +47,33 @@ public String getName() {
...
             } catch (Exceptions.YamlValidationException e) {
-                System.out.println("YAML file has schema error: " + e.getMessage());
+                System.out.println("YAML file \"" + e.configPath + "\" has schema error: " + e.getMessage());
                 return JavaSee.ExitStatus.CONFIG_FILE_SCHEMA_ERROR;
             } catch (FileNotFoundException e) {
                 System.err.println("YAML file is not found: " + e.getMessage());
                 return JavaSee.ExitStatus.CONFIG_FILE_NOT_FOUND;
             } catch (YAMLException e) {
-                System.out.println("YAML file has syntax error: " + e.getMessage());
+                System.err.println("YAML file has syntax error: " + e.getMessage());
```
Comment (ybiquitous) [inline src/main/java/com/github/sider/javasee/command/CheckCommand.java:72]: [suggestion] Can we add a file path also here?

```suggestion
                System.err.println("YAML file \"" + configPath() + "\" has syntax error: " + e.getMessage());
```

File: src/main/java/com/github/sider/javasee/command/CheckCommand.java [hunk 2]
```diff
@@ -47,30 +47,33 @@ public String getName() {
...
+                return JavaSee.ExitStatus.CONFIG_FILE_NOT_FOUND;
             }
             var rootPath = Optional.ofNullable(optionRoot).map((root) -> new File(root)).orElse(configPath().getParentFile());
             Map<String, Object> yaml;
             Config config = null;
             try {
                 yaml = new Yaml().load(new FileInputStream(configPath()));
                 if(yaml == null) {
-                    System.out.println("YAML file has unknown error");
+                    System.out.println("YAML file \"" + configPath() + "\" has unknown error");
```
Comment (ybiquitous) [inline src/main/java/com/github/sider/javasee/command/CheckCommand.java:61]: [suggestion] Can we output the all error messages to STDERR like this?

```suggestion
                    err.println("YAML file \"" + configPath() + "\" has unknown error");
```

Comment (kmizu) [reply src/main/java/com/github/sider/javasee/command/CheckCommand.java]: I've fixed them. @ybiquitous

Comment (ybiquitous) [reply src/main/java/com/github/sider/javasee/command/CheckCommand.java]: Thanks!

File: src/main/java/com/github/sider/javasee/command/TestCommand.java [hunk 3]
```diff
@@ -51,28 +51,31 @@ public String getName() {
...
-            System.out.println("YAML file has schema error: " + e.getMessage());
+            err.println("YAML file \"" + e.configPath + "\" has schema error: " + e.getMessage());
             return JavaSee.ExitStatus.CONFIG_FILE_SCHEMA_ERROR;
         } catch (FileNotFoundException e) {
-            System.err.println("YAML file is not found: " + e.getMessage());
+            err.println("YAML file is not found: " + e.getMessage());
             return JavaSee.ExitStatus.CONFIG_FILE_NOT_FOUND;
         } catch (YAMLException e) {
-            System.out.println("YAML file has syntax error: " + e.getMessage());
+            err.println("YAML file has syntax error: " + e.getMessage());
```
Comment (ybiquitous) [inline src/main/java/com/github/sider/javasee/command/TestCommand.java:74]: [suggestion] Could you fix this error message also?

```suggestion
            err.println("YAML file \"" + configPath + "\" has syntax error: " + e.getMessage());
```

File: src/main/java/com/github/sider/javasee/command/CheckCommand.java [hunk 4]
```diff
@@ -47,30 +47,33 @@ public String getName() {
         formatter.onStart();
 
         try {
-            if(!configPath().isFile()) {
+            if(!configPath().isFile() || !configPath().exists()) {
```
Comment (ybiquitous) [inline src/main/java/com/github/sider/javasee/command/CheckCommand.java:50]: [suggestion] How about replace the multiple `configPath()` method calls with a local variable like this?

```java
var configPath = configPath():
```

Comment (kmizu) [reply src/main/java/com/github/sider/javasee/command/CheckCommand.java]: OK! I've fixed them.

Comment (ybiquitous) [reply src/main/java/com/github/sider/javasee/command/CheckCommand.java]: Thank you! 👍

File: src/main/java/com/github/sider/javasee/Config.java [hunk 5]
```diff
@@ -17,24 +22,67 @@ public Config(List<Rule> rules, File rootDirectory) {
...
+            if(imported) {
+                message = configPath + " being imported is not found ...";
+            } else {
+                message =
+                    "There is nothing to test at " + configPath + " ...\n" +
+                    "Make a configuration and run test again!";
+            }
+            throw new Exceptions.ConfigFileException(configPath, message, JavaSee.ExitStatus.CONFIG_FILE_NOT_FOUND);
+        }
+        Map<String, Object> yaml = new Yaml().load(new FileInputStream(configPath));
```
Comment (ybiquitous) [inline src/main/java/com/github/sider/javasee/Config.java:37]: > __[PMD Java]__ Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter ([view](https://sider.review/gh/repos/171470727/pulls/104#issue-4552701))
>
> <details>
>
> | Rule | Ruleset | Priority |
> | ---- | ------- | -------- |
> | `AvoidFileStream` | Performance | 1 |
>
> References:
> - <https://pmd.github.io/pmd-6.26.0/pmd_rules_java_performance.html#avoidfilestream>
>
> You can [close this issue](https://sider.review/gh/repos/171470727/pulls/104#issue-4552701) if no need to fix it. [Learn more](https://help.sider.review/getting-started/working-with-issues).
>
> </details>

Does this issue make sense?

Comment (kmizu) [reply src/main/java/com/github/sider/javasee/Config.java:37]: It doesn't make issue in this context.  This issue make sense only in performance-sensitive contexts.

Comment (ybiquitous) [reply src/main/java/com/github/sider/javasee/Config.java:37]: I got it. Thanks. 👍

File: src/main/java/com/github/sider/javasee/command/TestCommand.java (hunk 3 above)
Comment (kmizu) [reply src/main/java/com/github/sider/javasee/command/TestCommand.java]: I've fixed it in https://github.com/sider/JavaSee/pull/104/commits/82da0e44c4c79fc30a3e5e9becaf36baaba7ea6a

**Review by ybiquitous — APPROVED** [review]
Thank you so much. I could confirm this feature works and I've closed all the Sider issues because they don't make sense.
(if you have an opposite idea about this, please let me know.)

🎉 🎉 🎉

**Comment by ybiquitous** [comment]
Could you please squash to one commit?

**Comment by kmizu** [comment]
> Could you please squash to one commit?

I've squashed right now.

**Pull request merged**